File: r565/E.cc
#include <iostream>
#include <vector>
#include <tuple>

using namespace std;

int main() {
    int t;
    cin >> t;
    while(t--) {
        int n, m;
        cin >> n >> m;
        vector<vector<int>> e(n);
        while(m--) {
            int u, v;
            cin >> u >> v;
            e[u-1].push_back(v-1);
            e[v-1].push_back(u-1);
        }

        vector<bool> seen(n);
        vector<int> s[2];
        vector<pair<int,int>> q(1, {0,0});
        while(q.size()) {
            int v, j;
            tie(v, j) = q.back();
            q.pop_back();
            if(seen[v]) continue; // been here
            seen[v] = true;
            s[j].push_back(v);
            for(int i : e[v]) q.push_back({i, j^1});
        }

        vector<int> & ans = (s[0].size() < s[1].size() ? s[0] : s[1]);
        cout << ans.size() << endl;
        for(int v : ans) cout << v+1 << " "; cout << endl;
    }
}

File: r565/E.py
t = int(input())

for i in range(t):
    n, m = map(int, input().split())
    e = [[] for i in range(n)]
    for j in range(m):
        u, v = map(int, input().split())
        e[u-1].append(v-1)
        e[v-1].append(u-1)
    q = [(0,0)] # make 0 vertex tree root
    s = [[], []] # two alternative sets
    seen = [False] * n # will become -1, 1
    while q: # depth first search to make a tree
        (v, j) = q.pop()
        if seen[v]: continue # been here
        seen[v] = True
        s[j].append(v)
        #for v in e[v]: q.append((v, j^1))
        q += ((v, j^1) for v in e[v])
    ans = s[0] if len(s[0]) < len(s[1]) else s[1]
    print(len(ans))
    print(' '.join(str(v+1) for v in ans))
